bot/cogs/moderation/incidents.py

```diff
@@ -0,0 +1,381 @@
+        The crawl assumes that the channel history doesn't change as we go over it.
+
+        Behaviour is configured by: `CRAWL_LIMIT`, `CRAWL_SLEEP`.
+        """
+        await self.bot.wait_until_guild_available()
+        incidents: discord.TextChannel = self.bot.get_channel(Channels.incidents)
+
+        log.debug(f"Crawling messages in #incidents: {CRAWL_LIMIT=}, {CRAWL_SLEEP=}")
+        async for message in incidents.history(limit=CRAWL_LIMIT):
+
+            if not is_incident(message):
+                log.trace(f"Skipping message {message.id}: not an incident")
+                continue
+
+            if has_signals(message):
+                log.trace(f"Skipping message {message.id}: already has all signals")
+                continue
+
+            await add_signals(message)
+            await asyncio.sleep(CRAWL_SLEEP)
+
+        log.debug("Crawl task finished!")
+
+    async def archive(self, incident: discord.Message, outcome: Signal, actioned_by: discord.Member) -> bool:
+        """
+        Relay an embed representation of `incident` to the #incidents-archive channel.
+
+        The following pieces of information are relayed:
+            * Incident message content (as embed description)
+            * Incident attachment (if image, shown in archive embed)
+            * Incident author name (as webhook author)
+            * Incident author avatar (as webhook avatar)
+            * Resolution signal `outcome` (as embed colour & footer)
+            * Moderator `actioned_by` (name & discriminator shown in footer)
+
+        If `incident` contains an attachment, we try to add it to the archive embed. There is
+        no handing of extensions / file types - we simply dispatch the attachment file with the
+        webhook, and try to display it in the embed. Testing indicates that if the attachment
+        cannot be displayed (e.g. a text file), it's invisible in the embed, with no error.
+
+        Return True if the relay finishes successfully. If anything goes wrong, meaning
+        not all information was relayed, return False. This signals that the original
+        message is not safe to be deleted, as we will lose some information.
+        """
+        log.debug(f"Archiving incident: {incident.id} (outcome: {outcome}, actioned by: {actioned_by})")
+        embed = make_embed(incident, outcome, actioned_by)
+
+        # If the incident had an attachment, we will try to relay it
+        if incident.attachments:
+            attachment = incident.attachments[0]  # User-sent messages can only contain one attachment
+            log.debug(f"Attempting to archive incident attachment: {attachment.filename}")
+
+            attachment_file = await attachment.to_file()  # The file will be sent with the webhook
+            embed.set_image(url=f"attachment://{attachment.filename}")  # Embed displays the attached file
+        else:
+            attachment_file = None
```

Comment:
Not clean, but as clean as I could get it. There's something oddly difficult here because we make the embed in `make_embed`, but there also has to be some branching here to determine whether a file should be passed to the `send`.

So if `make_embed` also looks at `incident.attachments` to determine whether `set_image` should be called, and this looks at `incident.attachments` to see whether a file should be passed, we're checking the same thing in two places and it feels ugly & fragile since they could easily go out of sync should one make adjustments to one method or the other.

Using this approach, we can use a single `if-else` to handle both the file and the url, and it feels ok, but it comes at the cost of still mutating the embed instance outside of `make_embed`.

Reply:
This is perfectly fine in my opinion.